admin/cleanup.py

```diff
@@ -0,0 +1,379 @@
+    if date_string is None:
+        return None
+    else:
+        return parse_date(date_string)
+
+
+def get_rackspace_driver(rackspace):
+    """
+    Get a libcloud Rackspace driver given some credentials and other
+    configuration.
+    """
+    rackspace = get_driver(Provider.RACKSPACE)(
+        rackspace['username'], rackspace['key'],
+        region=rackspace['region'],
+    )
+    return rackspace
+
+
+def get_ec2_driver(aws):
+    """
+    Get a libcloud EC2 driver given some credentials and other configuration.
+    """
+    ec2 = get_driver(Provider.EC2)(
+        aws['access_key'], aws['secret_access_token'],
+        region=aws['region'],
+    )
+    return ec2
+
+
+def _get_volume_creation_time(volume):
+    """
+    Extract the creation time from an AWS or Rackspace volume.
+
+    libcloud doesn't represent volume creation time uniformly across
+    drivers.  Thus this method only works on drivers specifically accounted
+    for.
+    """
+    try:
+        # AWS
+        return volume.extra['create_time']
+    except KeyError:
+        # Rackspace.  Timestamps have no timezone indicated.  Manual
+        # experimentation indicates timestamps are in UTC (which of course
+        # is the only reasonable thing).
+        return parse_date(
+            volume.extra['created_at']
+        ).replace(tzinfo=tzutc())
+
+
+def _get_volume_region(volume):
+    """
+    Get the name of the region the volume is in.
+    """
+    return (
+        # Rackspace
+        getattr(volume.driver, "region", None) or
+        # AWS
+        getattr(volume.driver, "region_name", None)
+    )
+
+
+def _describe_volume(volume):
+    """
+    Create a dictionary giving lots of interesting details about a cloud
+    volume.
+    """
+    return {
+        'id': volume.id,
+        'creation_time': _format_time(
+            _get_volume_creation_time(volume),
+        ),
+        'provider': volume.driver.name,
+        'region': _get_volume_region(volume),
+        # *Stuffed* with non-JSON-encodable goodies.
+        'extra': repr(volume.extra),
+    }
+
+
+@attributes(["lag", "marker"])
+class CleanVolumes(object):
+    """
+    Destroy volumes that leaked into the cloud from the acceptance and
+    functional test suites.
+    """
+    def start(self, config, dry_run=False):
+        """
+        Clean up old volumes belonging to test-created Flocker clusters.
+        """
+        drivers = self._get_cloud_drivers(config)
+        volumes = self._get_cloud_volumes(drivers)
+        actions = self._filter_test_volumes(self.lag, volumes)
+        if not dry_run:
+            self._destroy_cloud_volumes(actions.destroy)
+        return actions
+
+    def _get_cloud_drivers(self, config):
+        """
+        From the private buildbot configuration, construct a list of all of the
+        libcloud drivers where leaked volumes might be found.
+        """
+        base_ec2 = config["aws"]
```

Comment:
nit: inline below just like for rackspace. Better yet, leverage https://github.com/ClusterHQ/flocker/blob/master/flocker/provision/__init__.py#L16 somehow so you don't have to explicitly list all supported clouds here as well.